### PR TITLE
[BOLT][AArch64] Don't change layout in PatchEntries

### LINKED ISSUE #70771: BOLT fails to instrument AArch64 trunk Clang linked with mold
AArch64 buildbot https://lab.llvm.org/buildbot/#/builders/276 is applying BOLT to Clang with Clang-BOLT CMake automation using trunk versions of both.

It has been failing with the following assertion when instrumenting mold-linked Clang:
```
BOLT-ERROR: JITLink failed: In graph in-memory object file, section .text: relocation target "_ZN5clang19RecursiveASTVisitorINS_5arcmt5trans13BodyTransformIN12_GLOBAL__N_118UnusedInitRewriterEEEE12TraverseTypeENS_8QualTypeE/1" at address 0x571bce8 is out of range of Branch26PCRel fixup at 0xeab9b88 (__hot_start, 0x7c00000 + 0x6eb9b88)
```

To repro: 
```
cmake ../llvm-project/llvm -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DLLVM_APPEND_VC_REV=OFF -DCMAKE_C_COMPILER_LAUNCHER=ccache -DCMAKE_CXX_COMPILER_LAUNCHER=ccache -DLLVM_USE_LINKER=mold '-DLLVM_ENABLE_PROJECTS=bolt;llvm;clang' -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_ASSERTIONS=ON '-DLLVM_LIT_ARGS=-v -vv' -GNinja -C../llvm-project/clang/cmake/caches/BOLT.cmake
```

Note that it passes if Clang is linked with lld.

### PR DESCRIPTION
**Pull request by yota9**
Due to LongJmp pass that is executed before PatchEntries we can't ignore
the function here since it would change pre-calculated output layout.
The test reloc-26 relied on the wrong behavior, rewritten to unittest.
This is also attemp to fix #70771

### CODE REVIEW

File: bolt/lib/Passes/PatchEntries.cpp [hunk 1]
```diff
@@ -98,6 +98,17 @@ void PatchEntries::runOnFunctions(BinaryContext &BC) {
     });
 
     if (!Success) {
+      // We can't change output layout for AArch64 due to LongJmp pass
```
Comment (aaupov) [inline bolt/lib/Passes/PatchEntries.cpp:101]: Is it safe to just continue without ignoring the function? 
And in general, what's the interaction between PatchEntries and LongJmp passes? Can we patch the entry and rely on LongJmp to insert trampolines?

**Review by aaupov — COMMENTED** [review]
The patch seems to be addressing the issue but I don't know if it's safe in general.

File: bolt/lib/Passes/PatchEntries.cpp (hunk 1 above)
Comment (yota9) [reply bolt/lib/Passes/PatchEntries.cpp:101]: I would say yes. Internally we're using rewrite option so old text is completely removed, so we expect everything to be disassembled properly. To be honest I didn't even know that patch entries is forced to run. Also patch entries skips functions which size is smaller than 128 anyway, that we might consider not safe too, but we're doing it :) And finally the chance to call old function on ARM is smaller due to +- 128 mb limit.

**Comment by yota9** [comment]
@aaupov Any objections here? :)

**Comment by aaupov** [comment]
@yota9: Triggered internal testing, will get back after it's done.

File: bolt/unittests/Core/BinaryContext.cpp [hunk 2]
```diff
@@ -62,6 +62,45 @@ INSTANTIATE_TEST_SUITE_P(X86, BinaryContextTester,
 INSTANTIATE_TEST_SUITE_P(AArch64, BinaryContextTester,
                          ::testing::Values(Triple::aarch64));
 
+TEST_P(BinaryContextTester, FlushPendingRelocCALL26) {
+  if (GetParam() != Triple::aarch64)
+    GTEST_SKIP();
+
+  // This test checks that encodeValueAArch64 used by flushPendingRelocations
+  // returns right encoded values for CALL26 relocation for both backward and
```
Comment (aaupov) [inline bolt/unittests/Core/BinaryContext.cpp:70]: >  // returns correctly encoded values for CALL26 relocation for both backward and

**Review by aaupov — APPROVED** [review]
Testing has returned green. I'm still a bit concerned about the interaction b/w PatchEntries and LongJmp, but I guess we can live with that unless we hit edge cases. AArch64 test with `patch-entries` would be helpful.
LGTM with one nit.

**Comment by yota9** [comment]
Thanks @aaupov . Well why not, since it is simple-enough add simple patch-entries patch, although it is a bit out of scope of this change :)

**Pull request merged**